File: sources/LiquidSense Release/me/aquavit/liquidsense/module/modules/blatant/AutoWeapon.java
package me.aquavit.liquidsense.module.modules.blatant;

import me.aquavit.liquidsense.utils.item.ItemUtils;
import me.aquavit.liquidsense.event.events.AttackEvent;
import me.aquavit.liquidsense.event.EventTarget;
import me.aquavit.liquidsense.event.events.PacketEvent;
import me.aquavit.liquidsense.event.events.UpdateEvent;
import me.aquavit.liquidsense.module.Module;
import me.aquavit.liquidsense.module.ModuleCategory;
import me.aquavit.liquidsense.module.ModuleInfo;
import me.aquavit.liquidsense.value.BoolValue;
import net.minecraft.enchantment.Enchantment;
import net.minecraft.entity.ai.attributes.AttributeModifier;
import net.minecraft.item.ItemStack;
import net.minecraft.item.ItemSword;
import net.minecraft.item.ItemTool;
import net.minecraft.network.play.client.C02PacketUseEntity;
import net.minecraft.network.play.client.C09PacketHeldItemChange;

@ModuleInfo(name = "AutoWeapon", description = "Automatically selects the best weapon in your hotbar.", category = ModuleCategory.BLATANT)
public class AutoWeapon extends Module {
    private BoolValue silentValue = new BoolValue("SpoofItem", false);

    private C02PacketUseEntity packetUseEntity;
    private boolean spoofedSlot = false;
    private boolean gotIt = false;
    private int tick;

    @EventTarget
    public void onAttack(AttackEvent event){
        gotIt = true;
    }

    @EventTarget
    public void onPacket(PacketEvent event){
        if (event.getPacket() instanceof C02PacketUseEntity && ((C02PacketUseEntity)event.getPacket()).getAction() == C02PacketUseEntity.Action.ATTACK && this.gotIt) {
            gotIt = false;

            int slot = -1;
            double bestDamage = 0.0;

            for (int i=0; i<=8; i++){
                ItemStack itemStack = mc.thePlayer.inventory.mainInventory[i];
                if (itemStack != null && (itemStack.getItem() instanceof ItemSword || itemStack.getItem() instanceof ItemTool)) {
                    for (AttributeModifier attributeModifier : itemStack.getAttributeModifiers().get("generic.attackDamage")){
                        double damage = attributeModifier.getAmount() + 1.25 * (double) ItemUtils.getEnchantment(itemStack, Enchantment.sharpness);

                        if (damage > bestDamage) {
                            bestDamage = damage;
                            slot = i;
                        }
                    }
                }
            }

            if (slot != -1 && slot != mc.thePlayer.inventory.currentItem) {
                if (silentValue.get()) {
                    mc.getNetHandler().addToSendQueue(new C09PacketHeldItemChange(slot));
                    spoofedSlot = true;
                } else {
                    mc.thePlayer.inventory.currentItem = slot;
                    mc.playerController.updateController();
                }

                event.cancelEvent();
                packetUseEntity = (C02PacketUseEntity) event.getPacket();
                tick = 0;
            }
        }
    }

    @EventTarget
    public void onUpdate(UpdateEvent event){
        if (tick < 1) {
            tick++;
            return;
        }

        if (packetUseEntity != null) {
            mc.getNetHandler().getNetworkManager().sendPacket(packetUseEntity);

            if (spoofedSlot)
                mc.getNetHandler().addToSendQueue(new C09PacketHeldItemChange(mc.thePlayer.inventory.currentItem));

            packetUseEntity = null;
        }
    }

}

File: sources/Kevin Reborn/kevin/module/modules/combat/AutoWeapon.kt
/*
 * This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package kevin.module.modules.combat

import kevin.event.AttackEvent
import kevin.event.EventTarget
import kevin.event.PacketEvent
import kevin.event.UpdateEvent
import kevin.module.*
import kevin.utils.ItemUtils
import net.minecraft.enchantment.Enchantment
import net.minecraft.item.ItemSword
import net.minecraft.item.ItemTool
import net.minecraft.network.play.client.C02PacketUseEntity
import net.minecraft.network.play.client.C09PacketHeldItemChange

class AutoWeapon : Module("AutoWeapon", "Automatically selects the best weapon in your hotbar.", category = ModuleCategory.COMBAT) {
    private val mode = ListValue("Mode", arrayOf("Direct", "Spoof"), "Direct")
    private val silentValue : Boolean
        get() = mode equal "Spoof"
    private val switchBack = BooleanValue("SwitchBack", true)
    private val ticksValue = IntegerValue("WaitTicks", 10, 1, 20)
    private val swordsFirst = BooleanValue("SwordsFirst",false)
    private var attackEnemy = false

    private var spoofedSlot = 0

    private var beforeSlot = -1

    @EventTarget
    fun onAttack(event: AttackEvent) {
        attackEnemy = true
    }

    @EventTarget
    fun onPacket(event: PacketEvent) {
        if (event.packet !is C02PacketUseEntity)
            return

        val thePlayer = mc.thePlayer ?: return

        val packet = event.packet

        if (packet.action == C02PacketUseEntity.Action.ATTACK && attackEnemy) {
            attackEnemy = false

            // Find the best weapon in hotbar (#Kotlin Style)
            var (slot, _) = (0..8)
                .map { Pair(it, thePlayer.inventory.getStackInSlot(it)) }
                .filter { it.second != null && (it.second?.item is ItemSword || it.second?.item is ItemTool) }
                .maxByOrNull {
                    it.second!!.attributeModifiers["generic.attackDamage"].first().amount + 1.25 * ItemUtils.getEnchantment(
                        it.second,
                        Enchantment.sharpness
                    )
                } ?: return

            if (swordsFirst.get()){
                val bestSword = (0..8)
                    .map { Pair(it, thePlayer.inventory.getStackInSlot(it)) }
                    .filter { it.second != null && it.second?.item is ItemSword }
                    .maxByOrNull {
                        it.second!!.attributeModifiers["generic.attackDamage"].first().amount + 1.25 * ItemUtils.getEnchantment(
                            it.second,
                            Enchantment.sharpness
                        )
                    }
                if (bestSword!=null){
                    slot = bestSword.first
                }
            }

            beforeSlot = thePlayer.inventory.currentItem
            if (switchBack.get()) spoofedSlot = ticksValue.get()

            if (slot == thePlayer.inventory.currentItem) // If in hand no need to swap
                return

            // Switch to best weapon
            if (silentValue) {
                mc.netHandler.addToSendQueue(C09PacketHeldItemChange(slot))
                spoofedSlot = ticksValue.get()
            } else {
                thePlayer.inventory.currentItem = slot
                mc.playerController.updateController()
            }

            // Resend attack packet
            mc.netHandler.addToSendQueue(packet)
            event.cancelEvent()
        }
    }

    @EventTarget
    fun onUpdate(update: UpdateEvent) {
        // Switch back to old item after some time
        if (spoofedSlot > 0) {
            if (spoofedSlot == 1)
                if (silentValue) {
                    mc.netHandler.addToSendQueue(C09PacketHeldItemChange(mc.thePlayer!!.inventory.currentItem))
                } else {
                    mc.thePlayer.inventory.currentItem = beforeSlot
                    mc.playerController.updateController()
                }
            spoofedSlot--
        }
    }
}